// Code generated by ent, DO NOT EDIT.

package ent

import (
	"context"
	"notionboy/db/ent/predicate"
	"notionboy/db/ent/quota"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// QuotaDelete is the builder for deleting a Quota entity.
type QuotaDelete struct {
	config
	hooks    []Hook
	mutation *QuotaMutation
}

// Where appends a list predicates to the QuotaDelete builder.
func (qd *QuotaDelete) Where(ps ...predicate.Quota) *QuotaDelete {
	qd.mutation.Where(ps...)
	return qd
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (qd *QuotaDelete) Exec(ctx context.Context) (int, error) {
	return withHooks[int, QuotaMutation](ctx, qd.sqlExec, qd.mutation, qd.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (qd *QuotaDelete) ExecX(ctx context.Context) int {
	n, err := qd.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (qd *QuotaDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(quota.Table, sqlgraph.NewFieldSpec(quota.FieldID, field.TypeInt))
	if ps := qd.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, qd.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	qd.mutation.done = true
	return affected, err
}

// QuotaDeleteOne is the builder for deleting a single Quota entity.
type QuotaDeleteOne struct {
	qd *QuotaDelete
}

// Where appends a list predicates to the QuotaDelete builder.
func (qdo *QuotaDeleteOne) Where(ps ...predicate.Quota) *QuotaDeleteOne {
	qdo.qd.mutation.Where(ps...)
	return qdo
}

// Exec executes the deletion query.
func (qdo *QuotaDeleteOne) Exec(ctx context.Context) error {
	n, err := qdo.qd.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{quota.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (qdo *QuotaDeleteOne) ExecX(ctx context.Context) {
	if err := qdo.Exec(ctx); err != nil {
		panic(err)
	}
}
